component/prometheus/source/api/api_test.go

```diff
@@ -0,0 +1,392 @@
+				serverExit <- c.Run(ctx)
+			}()
+
+			comp := c.(*Component)
+			waitForServerToBeReady(t, comp.args)
+
+			initialServer := comp.server
+			require.NotNil(t, initialServer)
+
+			err = c.Update(tc.newArgs)
+			require.NoError(t, err)
+
+			waitForServerToBeReady(t, comp.args)
+
+			require.NotNil(t, comp.server)
+			restarted := initialServer != comp.server
+
+			require.Equal(t, tc.shouldRestart, restarted)
+
+			// shut down cleanly to release ports for other tests
+			cancel()
+			select {
+			case err := <-serverExit:
+				require.NoError(t, err, "unexpected error on server exit")
+			case <-time.After(5 * time.Second):
+				t.Fatalf("timed out waiting for server to shut down")
+			}
+		})
+	}
+}
+
+type testSample struct {
+	ts  int64
+	val float64
+	l   labels.Labels
+}
+
+func waitForServerToBeReady(t *testing.T, args Arguments) {
+	require.Eventuallyf(t, func() bool {
+		resp, err := http.Get(fmt.Sprintf(
+			"http://%v:%d/wrong/path",
+			args.Server.HTTP.ListenAddress,
+			args.Server.HTTP.ListenPort,
+		))
+		t.Logf("err: %v, resp: %v", err, resp)
+		return err == nil && resp.StatusCode == 404
+	}, 5*time.Second, 20*time.Millisecond, "server failed to start before timeout")
+}
+
+func verifyExpectations(
+	t *testing.T,
+	input []prompb.TimeSeries,
+	expected []testSample,
+	actualSamples chan testSample,
+	args Arguments,
+	ctx context.Context,
+) {
+	// In case server didn't start yet
+	waitForServerToBeReady(t, args)
+
+	// Send the input time series to the component
+	endpoint := fmt.Sprintf(
+		"http://%s:%d/api/v1/metrics/write",
+		args.Server.HTTP.ListenAddress,
+		args.Server.HTTP.ListenPort,
+	)
+	err := request(ctx, endpoint, &prompb.WriteRequest{Timeseries: input})
+	require.NoError(t, err)
+
+	// Verify we receive expected metrics
+	for _, exp := range expected {
+		select {
+		case actual := <-actualSamples:
+			require.Equal(t, exp, actual)
+		case <-ctx.Done():
+			t.Fatalf("test timed out")
+		}
+	}
+
+	select {
+	case unexpected := <-actualSamples:
+		t.Fatalf("unexpected extra sample received: %v", unexpected)
+	default:
+	}
+}
+
+func testAppendable(actualSamples chan testSample) []storage.Appendable {
+	hookFn := func(
+		ref storage.SeriesRef,
+		l labels.Labels,
+		ts int64,
+		val float64,
+		next storage.Appender,
+	) (storage.SeriesRef, error) {
+
+		actualSamples <- testSample{ts: ts, val: val, l: l}
+		return ref, nil
+	}
+
+	return []storage.Appendable{agentprom.NewInterceptor(
+		nil,
+		agentprom.WithAppendHook(
+			hookFn))}
+}
+
+func request(ctx context.Context, rawRemoteWriteURL string, req *prompb.WriteRequest) error {
+	remoteWriteURL, err := url.Parse(rawRemoteWriteURL)
+	if err != nil {
+		return err
+	}
+
+	client, err := remote.NewWriteClient("remote-write-client", &remote.ClientConfig{
+		URL:     &config.URL{URL: remoteWriteURL},
+		Timeout: model.Duration(30 * time.Second),
+	})
+	if err != nil {
+		return err
+	}
+
+	buf, err := proto.Marshal(protoadapt.MessageV2Of(req))
```

Comment:
We need to use `protoadapt.MessageV2Of` since the Agent repo uses the new V2 [google.golang.org/protobuf](https://pkg.go.dev/mod/google.golang.org/protobuf) library, while prometheus uses the V1 [github.com/golang/protobuf](https://pkg.go.dev/mod/github.com/golang/protobuf) library. Importing the V1 is a linter error.

In order to use the `protoadapt`, I had to upgrade the `google.golang.org/protobuf` to 1.30.0 - see `go.mod`.

I don't think it's a problem since it's a minor version and we would eventually want to update this dependency anyway.